### ISSUE TITLE
SmartScraperMultiConcatGraph error with bedrock

### ISSUE THREAD

**Issue by rjbks** [ScrapeGraphAI/Scrapegraph-ai#687]
Setup:
```python
In [67]: client = boto3.client("bedrock-runtime", region_name="us-west-2")
In [68]: graph_config = {
    ...:     "llm": {
    ...:         "client": client,
    ...:         "model": "bedrock/anthropic.claude-3-5-sonnet-20240620-v1:0",
    ...:         "temperature": 0.0
    ...:     },
    ...:     'format': 'json'
    ...: }
In [69]: g = SmartScraperMultiConcatGraph(
    ...:     prompt="Find information on all Fellowship programs offered, current and historic, including date ranges if applicable.",
    ...:     source= [
    ...:         "https://www.childrensdmc.org/health-professionals/just-for-doctors/fellowships/infectious-diseases",
    ...:         "https://www.medstarhealth.org/education/fellowship-programs/infectious-disease"
    ...:         ],
    ...:     schema=None,
    ...:     config=graph_config
    ...: )
    ...: 
    ...: result = g.run()
--- Executing GraphIterator Node with batchsize 16 ---
...
/scrapegraphai/nodes/graph_iterator_node.py:106, in GraphIteratorNode._async_execute(self, state, batchsize)
    103 if graph_instance is None:
    104     raise ValueError("graph instance is required for concurrent execution")
--> 106 graph_instance = [graph_instance(
    107     prompt="",
    108     source="",
    109     config=scraper_config,
    110     schema=self.schema) for _ in range(len(urls))]
    112 for graph in graph_instance:
    113     if "graph_depth" in graph.config:

TypeError: 'SmartScraperGraph' object is not callable
```

**Comment by VinciGit00** [comment]
Ok which version of scrapegraph are you using?

**Comment by rjbks** [comment]
1.21.0

**Comment by VinciGit00** [comment]
does it work with normal smart scraper?

**Comment by rjbks** [comment]
Yes and no. It throws an error when not specifying a schema, but within the error is correct JSON:
```

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/langchain_core/output_parsers/json.py:89, in JsonOutputParser.parse_result(self, result, partial)
     87 except JSONDecodeError as e:
     88     msg = f"Invalid json output: {text}"
---> 89     raise OutputParserException(msg, llm_output=text) from e

OutputParserException: Invalid json output: Here is the JSON response based on the website content:

{
  "fellowship_programs": [
    {
      "name": "Women's Health Fellowship",
      "link": "https://siouxfallsfpr.org/residency-experience/womens-health-fellowship/"
    }
  ],
  "residency_programs": [
    {
      "name": "Family Medicine Residency",
      "description": "3-year program with curriculum for each year",
      "links": [
        "https://siouxfallsfpr.org/residency-experience/year-1-curriculum-benefits/",
        "https://siouxfallsfpr.org/residency-experience/year-2-curriculum-benefits/", 
        "https://siouxfallsfpr.org/residency-experience/year-3-curriculum-benefits/"
      ]
    },
    {
      "name": "Rural Track",
      "location": "Pierre, SD",
      "link": "https://pierreruralfm.org"
    }
  ],
  "additional_info": "The program offers both allopathic and osteopathic approaches, with a focus on training physicians for rural communities in South Dakota and the Upper Midwest. Residents can customize their experience through electives and rotations."
}
```

I have not tried with schema yet.

**Comment by VinciGit00** [comment]
thx I understand the problem and in the next hours I will try to fix it

**Comment by VinciGit00** [comment]
ok please try to update to the new beta

**Comment by rjbks** [comment]
I believe this has to do with passing in an instantiated client in the
client key.

**Comment by rjbks** [comment]
@VinciGit00 There seem to be 2 issues here, the op one I reported with multiconcatgraph, that one still exists in the new release ( scrapegraphai-1.21.2b1) with the same traceback: 
```python
In [7]: client = boto3.client("bedrock-runtime", region_name="us-west-2")

In [8]: graph_config = {
   ...:     ...:     "llm": {
   ...:     ...:         "client": client,
   ...:     ...:         "model": "bedrock/anthropic.claude-3-5-sonnet-20240620-v1:0",
   ...:     ...:         "temperature": 0.0
   ...:     ...:     },
   ...:     ...:     'format': 'json'
   ...:     ...: }

In [9]: g = SmartScraperMultiConcatGraph(
   ...:     ...:     prompt="Find information on all Fellowship programs offered, current and historic, including date ranges if applicable.",
   ...:     ...:     source= [
   ...:     ...:         "https://www.childrensdmc.org/health-professionals/just-for-doctors/fellowships/infectious-diseases",
   ...:     ...:         "https://www.medstarhealth.org/education/fellowship-programs/infectious-disease"
   ...:     ...:         ],
   ...:     ...:     schema=None,
   ...:     ...:     config=graph_config
   ...:     ...: )
In [10]: g.run()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[11], line 1
----> 1 g.run()

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/smart_scraper_multi_concat_graph.py:103, in SmartScraperMultiConcatGraph.run(self)
     96 """
     97 Executes the web scraping and searching process.
     98 
     99 Returns:
    100     str: The answer to the prompt.
    101 """
    102 inputs = {"user_prompt": self.prompt, "urls": self.source}
--> 103 self.final_state, self.execution_info = self.graph.execute(inputs)
    105 return self.final_state.get("answer", "No answer found.")

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/base_graph.py:258, in BaseGraph.execute(self, initial_state)
    256     return (result["_state"], [])
    257 else:
--> 258     return self._execute_standard(initial_state)

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/base_graph.py:179, in BaseGraph._execute_standard(self, initial_state)
    166     graph_execution_time = time.time() - start_time
    167     log_graph_execution(
    168         graph_name=self.graph_name,
    169         source=source,
   (...)
    177         exception=str(e)
    178     )
--> 179     raise e
    180 node_exec_time = time.time() - curr_time
    181 total_exec_time += node_exec_time

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/base_graph.py:163, in BaseGraph._execute_standard(self, initial_state)
    161 with self.callback_manager.exclusive_get_callback(llm_model, llm_model_name) as cb:
    162     try:
--> 163         result = current_node.execute(state)
    164     except Exception as e:
    165         error_node = current_node.node_name

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/nodes/graph_iterator_node.py:73, in GraphIteratorNode.execute(self, state)
     71     state = eventloop.run_until_complete(self._async_execute(state, batchsize))
     72 else:
---> 73     state = asyncio.run(self._async_execute(state, batchsize))
     75 return state

File /opt/anaconda3/envs/med_device/lib/python3.12/asyncio/runners.py:194, in run(main, debug, loop_factory)
    190     raise RuntimeError(
    191         "asyncio.run() cannot be called from a running event loop")
    193 with Runner(debug=debug, loop_factory=loop_factory) as runner:
--> 194     return runner.run(main)

File /opt/anaconda3/envs/med_device/lib/python3.12/asyncio/runners.py:118, in Runner.run(self, coro, context)
    116 self._interrupt_count = 0
    117 try:
--> 118     return self._loop.run_until_complete(task)
    119 except exceptions.CancelledError:
    120     if self._interrupt_count > 0:

File /opt/anaconda3/envs/med_device/lib/python3.12/asyncio/base_events.py:687, in BaseEventLoop.run_until_complete(self, future)
    684 if not future.done():
    685     raise RuntimeError('Event loop stopped before Future completed.')
--> 687 return future.result()

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/nodes/graph_iterator_node.py:106, in GraphIteratorNode._async_execute(self, state, batchsize)
    103 if graph_instance is None:
    104     raise ValueError("graph instance is required for concurrent execution")
--> 106 graph_instance = [graph_instance(
    107     prompt="",
    108     source="",
    109     config=scraper_config,
    110     schema=self.schema) for _ in range(len(urls))]
    112 for graph in graph_instance:
    113     if "graph_depth" in graph.config:

TypeError: 'SmartScraperGraph' object is not callable
```
I assume this has to do with a recent refactor (I forget the release) which expects an uninstantiated graph instance as the client value in the config, whereas this is getting an instantiated boto3 client.

Then there is the SmartScraperGraph JSON formatting issue. which now throws this after the update:

```python
In [8]: graph_config = {
   ...:     ...:     "llm": {
   ...:     ...:         "client": client,
   ...:     ...:         "model": "bedrock/anthropic.claude-3-5-sonnet-20240620-v1:0",
   ...:     ...:         "temperature": 0.0
   ...:     ...:     },
   ...:     ...:     'format': 'json'
   ...:     ...: }
In [14]: g = SmartScraperGraph(
    ...:        prompt="Find information on all Fellowship programs offered, current and historic, including date ranges if applicable.",
    ...:        source= "https://www.medstarhealth.org/education/fellowship-programs/infectious-disease",
    ...:        schema=None,
    ...:        config=graph_config
    ...:     )

In [15]: g.run()
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
Cell In[15], line 1
----> 1 g.run()

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/smart_scraper_graph.py:114, in SmartScraperGraph.run(self)
    106 """
    107 Executes the scraping process and returns the answer to the prompt.
    108 
    109 Returns:
    110     str: The answer to the prompt.
    111 """
    113 inputs = {"user_prompt": self.prompt, self.input_key: self.source}
--> 114 self.final_state, self.execution_info = self.graph.execute(inputs)
    116 return self.final_state.get("answer", "No answer found.")

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/base_graph.py:258, in BaseGraph.execute(self, initial_state)
    256     return (result["_state"], [])
    257 else:
--> 258     return self._execute_standard(initial_state)

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/base_graph.py:179, in BaseGraph._execute_standard(self, initial_state)
    166     graph_execution_time = time.time() - start_time
    167     log_graph_execution(
    168         graph_name=self.graph_name,
    169         source=source,
   (...)
    177         exception=str(e)
    178     )
--> 179     raise e
    180 node_exec_time = time.time() - curr_time
    181 total_exec_time += node_exec_time

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/graphs/base_graph.py:163, in BaseGraph._execute_standard(self, initial_state)
    161 with self.callback_manager.exclusive_get_callback(llm_model, llm_model_name) as cb:
    162     try:
--> 163         result = current_node.execute(state)
    164     except Exception as e:
    165         error_node = current_node.node_name

File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/nodes/generate_answer_node.py:131, in GenerateAnswerNode.execute(self, state)
    124     template_merge_prompt  = self.additional_info + template_merge_prompt
    126 if len(doc) == 1:
    127     prompt = PromptTemplate(
    128         template=template_no_chunks_prompt ,
    129         input_variables=["question"],
    130         partial_variables={"context": doc,
--> 131                             "format_instructions": format_instructions})
    132     chain =  prompt | self.llm_model | output_parser
    133     answer = chain.invoke({"question": user_prompt})

UnboundLocalError: cannot access local variable 'format_instructions' where it is not associated with a value
```

**Comment by VinciGit00** [comment]
ok, I can fix the second issue, what about he first one?

**Comment by rjbks** [comment]
@VinciGit00 This is the first issue (unresolved)
```
File /opt/anaconda3/envs/med_device/lib/python3.12/site-packages/scrapegraphai/nodes/graph_iterator_node.py:106, in GraphIteratorNode._async_execute(self, state, batchsize)
    103 if graph_instance is None:
    104     raise ValueError("graph instance is required for concurrent execution")
--> 106 graph_instance = [graph_instance(
    107     prompt="",
    108     source="",
    109     config=scraper_config,
    110     schema=self.schema) for _ in range(len(urls))]
    112 for graph in graph_instance:
    113     if "graph_depth" in graph.config:

TypeError: 'SmartScraperGraph' object is not callable
```

It has to do with instantiating the GraphIteratorNode with an already instantiated smart scraper graph node. Here is how the functioning SmartScraperMultiGraph code works (_create_graph:71):
```python
graph_iterator_node = GraphIteratorNode(
            input="user_prompt & urls",
            output=["results"],
            node_config={
                "graph_instance": SmartScraperGraph,
                "scraper_config": self.copy_config,
            },
            schema=self.copy_schema
        )
```

Here is how SmartScraperMultiConcatGraph (the one causing that error) does it (_create_graph:63):
```python
smart_scraper_instance = SmartScraperGraph(
            prompt="",
            source="",
            config=self.copy_config,
            schema=self.copy_schema
        )

        graph_iterator_node = GraphIteratorNode(
            input="user_prompt & urls",
            output=["results"],
            node_config={
                "graph_instance": smart_scraper_instance,
            }
        )
```
It is passing in an instantiated SmartScraperGraph in the "graph_instance" key of the GraphIteratorNode config, where it is expecting an uninstantiated class.

**Comment by rjbks** [comment]
I have updated the SmartScraperMultiConcatGraph code to look like this:
```python
graph_iterator_node = GraphIteratorNode(
            input="user_prompt & urls",
            output=["results"],
            node_config={
                "graph_instance": SmartScraperGraph,
                "scraper_config": self.copy_config,
                "scraper_schema": self.copy_schema,
            }
        )
```
And now only left with the parsing error we spoke of earlier:
```
UnboundLocalError: cannot access local variable 'format_instructions' where it is not associated with a value
```

**Comment by rjbks** [comment]
As for the JSON parsing error, in `GenerateAnswerNode.execute` both `format_instructions` and `output_parser` do not have conditional branches instantiating those variables in the event the client is BedrockChat.

**Comment by VinciGit00** [comment]
hi please update to the new beta

**Comment by rjbks** [comment]
@VinciGit00 

Even with a schema and the new fix (v1.21.2b2), the schema is not respected. Getting valid JSON involves "asking the LLM" in the user prompt. 
Looks like in generate answer node line 56, the bedrock client always resolves to the last `else` statement, where `output_parser` and `format_instructions` are `None` and `""` respectively:
```python
        if self.node_config.get("schema", None) is not None:
            if isinstance(self.llm_model, (ChatOpenAI, ChatMistralAI)):
                self.llm_model = self.llm_model.with_structured_output(
                    schema=self.node_config["schema"]
                )
                output_parser = get_structured_output_parser(self.node_config["schema"])
                format_instructions = "NA"
            else:
                if not isinstance(self.llm_model, ChatBedrock):
                    output_parser = get_pydantic_output_parser(self.node_config["schema"])
                    format_instructions = output_parser.get_format_instructions()
                else:
                    output_parser = None
                    format_instructions = ""
        else:
            if not isinstance(self.llm_model, ChatBedrock):
                output_parser = JsonOutputParser()
                format_instructions = output_parser.get_format_instructions()
            else:
                output_parser = None
                format_instructions = ""
```
In GenerateAnswerNode.execute, changing this on line 45:
```python
       if self.node_config.get("schema", None) is not None:
            if isinstance(self.llm_model, (ChatOpenAI, ChatMistralAI)):
                self.llm_model = self.llm_model.with_structured_output(
                    schema=self.node_config["schema"]
                )
                output_parser = get_structured_output_parser(self.node_config["schema"])
                format_instructions = "NA"
            else:
                output_parser = get_pydantic_output_parser(self.node_config["schema"])
                format_instructions = output_parser.get_format_instructions()
```
Plus altering the prompt with additional instruction for JSON output, provided json correctly:
```python
{'products': {'item_1': {'program_links': None, 'programs': []}, 'item_2': {'program_links': None, 'programs': [{'link': 'https://www.medstarhealth.org/education/fellowship-programs/infectious-disease', 'type': 'fellowship', 'specialty': 'Infectious Disease', 'sub_specialty': None, 'name': 'Infectious Disease Fellowship Program', 'institution_name': 'MedStar Washington Hospital Center', 'institution_link': 'https://www.medstarhealth.org/', 'institution_address': 'Washington, D.C.', 'director': 'Saumil Doshi', 'director_phone': '(202) 877-7164', 'director_email': 'saumil.doshi@medstar.net'}]}}}
```
This is not ideal, as the idea here is not to need on rely on prompt tuning to ensure JSON output. Is there a reason why we test for `if not isinstance(self.llm_model, ChatBedrock):` when deciding on output parsers?
